### DISCUSSION TITLE
【質問】base3/filetest1 ビルドについて

### DISCUSSION THREAD

**Discussion by tsukada-shigeki** [toppers/users-forum#71] [TOPPERS Seminar]
竹内様
お世話になります。
１）新基礎３篇:２日目　１０１Pに関してお尋ねします。
２）ビデオでは、base3/fileteszt1に移動して、ビルドを行いfilecmd.exe作成と有りますが(1002参照ください）、stdint.h　がないエラー（1001参照ください）が発生します。
３）原因と対策をご教示願います。
塚田
![スクリーンショット (1002)](https://user-images.githubusercontent.com/106538497/173192965-75e9f300-8cfe-42ed-bf4e-f9451c86875d.png)
![スクリーンショット (1001)](https://user-images.githubusercontent.com/106538497/173192976-f49fbe0b-1eff-4c26-883d-0c74bb7f1f65.png)

**Comment by Ryohsuke-Takeuchi** [comment]
教材の開発環境として、最新のMSYS環境で学習を行って頂くようにお願いしております。
stdint.hは古いGCCでは、サポートしていません。cygwin、MSYSとも、version6以上のコンパイラではサポートしています。
まず、gcc --versionで、コンパイラのバージョンを確認してください。
![image](https://user-images.githubusercontent.com/91311953/173281131-0c4e39bd-00f4-4836-a4a9-89b882242a8b.png)
![image](https://user-images.githubusercontent.com/91311953/173281262-543d11d4-e472-4f18-9807-6e4ae2eeb6ae.png)

古いコンパイラならば、以下のように、stdint.hの参照をやめても、問題ありません。
![image](https://user-images.githubusercontent.com/91311953/173281810-09a733b0-cbee-47c6-9652-f541d65d5e1d.png)
![image](https://user-images.githubusercontent.com/91311953/173281956-e2fb6c18-f524-469b-81f6-7b75bc4b3d24.png)

できれば、gccで何か開発するのならば、バージョンを6以上にあげた方が良いかもしれません。

**Reply by tsukada-shigeki** [reply]
竹内様

お世話になります。
メール拝受致しました。早速のアドバイス有難うございました

１．ｇｃｃのバージョンは11.2.0でした(添付ファイルご確認願います）。
２．file_test.cをmy_program/sd_file1で問題なくビルドできました。
　　セミナーを進めます。
３．今回質問の件は古いカーネルライブラリ（libkernel.a）が関係していないでしょうか？
　　後日確認致します。
[gccversion.docx](https://github.com/toppers/users-forum/files/8888320/gccversion.docx)
以上宜しくお願い致します。
塚田

**Reply by Ryohsuke-Takeuchi** [reply]
stdint.hはfilecmd.c中のuint32_t型解決のために、インクルードしています。
stdint.hをインクルードせずに、uint32_t型がコンパイルエラーとならないのならば、gccのversion 11は
stdint.hを明示的に持たなくなったのかもしれません。（二次参照でsys/_stdint.hを参照しているのかも？）
filetest1は、GCCを使ってUNIX互換環境で、ビルドとテストをします。
libkernel.aは、arm-none-eabi-gcc（ARM用コンパイラ）で作成したライブラリなので、今回の件とは、関係ないと思います

**Reply by Ryohsuke-Takeuchi** [reply]
もし、unit32_t型がコンパイルエラーになるなら、unsigned int型に置き換えても問題ありません。64ビットコンパイラでも、unsigned intは32ビットの整数なので、サイズ的には問題ありません。

**Comment by Ryohsuke-Takeuchi** [comment]
塚田様のgccversion.docxを見ると、
９行目のエラー文が、#include_next <stdint.h>となっていますが、
file_cmd.cの9行目は　#include <stdint.h>のはずです。
９行目自体が、おかしいような気がすますが。

**Reply by Ryohsuke-Takeuchi** [reply]
間違えました。
10行目のsys/stat.hの参照からエラーになっています。
filecmd.cでは、stat.hを参照していませんので、10行目を削除してもらっても問題ありません。
sys/stat.hはC言語で状態を取り出すインクルードファイルなので、エラーになるのが、おかしいです。
コンパイラのバグっぽい雰囲気がします

**Reply by tsukada-shigeki** [reply]
竹内様
お世話になります。

早速のアドバイス有難うございます。
filecmdの中で　stdint.h　と　stat.h　をコメント化でOKになりました。
添付ファイルをご確認願います。
以上宜しくお願い致します。
塚田
[filecmd修正.zip](https://github.com/toppers/users-forum/files/8890883/filecmd.zip)
